titus-server-master/src/main/java/com/netflix/titus/master/loadbalancer/service/LoadBalancerEngine.java

```diff
@@ -163,33 +168,78 @@ public void shutdown() {
+                    if (!loadBalancers.isEmpty()) {
+                        logger.info("Task update in job associated to one or more load balancers, deregistering {} load balancer targets: {}",
+                                loadBalancers.size(), task.getJobId());
+                    }
+                    return updatesForLoadBalancers(loadBalancers, task, State.Deregistered);
+                });
     }
 
-    private Observable<LoadBalancerTarget> targetsForTrackedTasks(Observable<Task> tasks) {
-        return tasks.doOnNext(task -> logger.debug("Checking if task in job is being tracked: {}", task.getId()))
-                .map(task -> Pair.of(task, store.getAssociatedLoadBalancersSetForJob(task.getJobId())))
-                // A task with an empty set is not tracked by any load balancer
-                .filter(pair -> !pair.getRight().isEmpty())
-                .doOnNext(pair -> logger.info("Task update in job being tracked, enqueuing {} load balancer updates: {}",
-                        pair.getRight().size(), pair.getLeft()))
-                .flatMap(pair -> Observable.from(
-                        pair.getRight().stream().map(association -> new LoadBalancerTarget(
-                                association,
-                                pair.getLeft().getId(),
-                                pair.getLeft().getTaskContext().get(TaskAttributes.TASK_ATTRIBUTES_CONTAINER_IP)
-                        )).collect(Collectors.toList()))
-                );
+    private Observable<TargetStateBatchable> moveFromEvents(Observable<TaskUpdateEvent> taskMovedEvents) {
+        return taskMovedEvents.flatMapIterable(taskMoved -> {
+            ArrayList<TargetStateBatchable> changes = new ArrayList<>();
+            Task task = taskMoved.getCurrentTask();
+            String targetJobId = task.getJobId();
+            Set<JobLoadBalancer> targetJobLoadBalancers = store.getAssociatedLoadBalancersSetForJob(targetJobId);
+            String sourceJobId = task.getTaskContext().get(TaskAttributes.TASK_ATTRIBUTES_MOVED_FROM_JOB);
+            if (StringExt.isEmpty(sourceJobId)) {
+                logger.warn("Task moved to {} does not include the source job id: {}", task.getJobId(), task.getId());
+                return Collections.emptyList();
+            }
+            Set<JobLoadBalancer> sourceJobLoadBalancers = store.getAssociatedLoadBalancersSetForJob(sourceJobId);
+
+            // operations below will dedup load balancers present in both source and target Sets, so we avoid
+            // unnecessary churn by deregistering and registering the same targets on the same loadBalancerId
+
+            // register on load balancers associated with the target job
+            if (TaskHelpers.isStartedWithIp(task)) {
+                // targetJobLoadBalancers - sourceJobLoadBalancers
+                Set<JobLoadBalancer> toRegister = targetJobLoadBalancers.stream()
```

Comment:
The same filtering logic is used in two places. This could be extracted to a method and reused.

Reply:
I had it extracted before, but the two minor differences (the desired state, and the logging line) made it so extracting it ended in more lines of code than have it separate. I will give it another shot.

Reply:
nvm, I was thinking on a different part of the PR. I will extract this.